Basket API: Add option to set shipping and payment methods

tulimaki opened the pull request:
* Add service serializers for basket
* Add option to set payment and shipping methods through API
* Avoid having lazy gettext objects in the error messages

chessbr commented on shuup/core/api/basket.py:
```diff
@@ -506,12 +521,35 @@ def set_shipping_address(self, request, *args, **kwargs):
+    def _handle_setting_method(self, request, model, attr_field):
+        self.process_request()
+        serializer = MethodIDSerializer(data=request.data)
+        if serializer.is_valid():
```
you can invoke `serializer.is_valid(raise_exception=True)` so you don't need to return `return Response(serializers.errors, status=status.HTTP_400_BAD_REQUEST)` explicitly and as a bonus we remove a indentation :)

chessbr commented on shuup/core/api/basket.py:
```diff
@@ -506,12 +521,35 @@ def set_shipping_address(self, request, *args, **kwargs):
+        self.process_request()
+        serializer = MethodIDSerializer(data=request.data)
+        if serializer.is_valid():
+            if request.data.get("id"):
```
use the serializer here, you don't need to test this any more.. just use `serializer.validated_data['id']`

chessbr commented on shuup/core/api/service.py:
```diff
@@ -0,0 +1,42 @@
+    def get_price(self, service):
+        price = 0
+        request = self.context["request"]
+        if request.basket:
```
You are assuming that we have the basket on each request, which will be false when not using the Front middleware.
IMO, you should read the `source` from the context and whoever creates the instance of this serializer, needs to pass the `source` to the context to receive a valid price.

```python

source = self.context.get("source")
if source:
    price = service.get_total_cost(source).taxful_price.value
```

then when creating the serializer on a ViewSet
```python
class SomeViewSet(...):
    def get_serializer_context(self):
        context = super(SomeViewSet, self).get_serializer_context()
        context['source'] = self.request.basket
        return context
```

or even creating the serializer manually:
```python
PaymentMethodSerializer(..., context={..., 'source': order})
```

After doing this, you'll need to change the viewsets which uses these serializers

chessbr commented on shuup/core/basket/objects.py:
```diff
@@ -208,6 +208,38 @@ def billing_address(self, value):
+        self.shipping_method_id = (shipping_method.id if shipping_method else None)
+
+        if self.shipping_method_id and hasattr(self, "_data"):
+            self._load()['shipping_method_id'] = self.shipping_method_id
```
yeah, this is also at the old code, but this logic do not accept to clearing the shipping method right? this also happens with addresses and payment method. maybe we should accept `None` here?

chessbr commented on shuup_tests/api/test_basket_api_set_sevices.py:
```diff
@@ -0,0 +1,89 @@
+    method_data = response_data["shipping_method"]
+    assert method_data["id"] == shipping_method.id
+    assert method_data["translations"]["en"]["name"] == shipping_method.name
+    assert method_data["price"] == 0
```
it would be good to have tests which actually have prices

tulimaki commented:
Refactored setting and getting basket data a little. Now there is ways to unset addresses and methods. Tests should prove that this works. Also refactored the service serializer a little to the direction you suggested.

chessbr commented:
Awesome work!
